### PR TITLE
Loss masking for distillation

### PR DESCRIPTION
**Pull request by jlamypoirier**
# ✨ Description

Pass a loss mask to kwargs so we it can be used for distillation loss, aka cross-entropy from logits

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [x] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

### CODE REVIEW

**Comment by oleksost** [comment]
IN the documentation, [here](https://github.com/ServiceNow/Fast-LLM/blob/d7922afff977af0d50e97ca0b1009e21542c4371/docs/recipes/instruction-finetuning.md?plain=1#L143), the parameter `use_loss_masking_spans` should be under `batch`, right?

**Comment by jlamypoirier** [comment]
@oleksost Yes, looks like we forgot to update

File: fast_llm/functional/cross_entropy.py [hunk 1]
```diff
@@ -169,12 +178,15 @@ def cross_entropy_forward_backward(
...
         assert target.dtype.is_floating_point, target.dtype
+        if loss_mask is not None:
+            Assert.eq(loss_mask.shape, logits.shape[:-1])
     if group:
         Assert.eq(implementation, CrossEntropyImpl.fused)
-        return fused_cross_entropy_forward_backward(
+        return _fused_cross_entropy_forward_backward(
             logits, target, grad_output, logits_scale_factor, target_format, group
         )
     else:
```
Comment (oleksost) [inline fast_llm/functional/cross_entropy.py:192]: In return _CROSS_ENTROPY_IMPLEMENTATIONS[implementation](... ) we also need to pass loss mask?

File: fast_llm/functional/cross_entropy.py [hunk 2]
```diff
@@ -169,12 +178,15 @@ def cross_entropy_forward_backward(
...
+        assert loss_mask is None
     else:
         Assert.eq(target.shape, logits.shape)
         assert target.dtype.is_floating_point, target.dtype
+        if loss_mask is not None:
+            Assert.eq(loss_mask.shape, logits.shape[:-1])
     if group:
         Assert.eq(implementation, CrossEntropyImpl.fused)
-        return fused_cross_entropy_forward_backward(
+        return _fused_cross_entropy_forward_backward(
```
Comment (oleksost) [inline fast_llm/functional/cross_entropy.py:189]: Need to pass loss_mask?

File: fast_llm/functional/cross_entropy.py [hunk 3]
```diff
@@ -169,15 +178,18 @@ def cross_entropy_forward_backward(
     if target_format == TargetFormat.labels:
         Assert.eq(target.shape, logits.shape[:-1])
         Assert.eq(target.dtype, torch.int64)
+        assert loss_mask is None
     else:
         Assert.eq(target.shape, logits.shape)
         assert target.dtype.is_floating_point, target.dtype
+        if loss_mask is not None:
+            Assert.eq(loss_mask.shape, logits.shape[:-1])
```
Comment (oleksost) [inline fast_llm/functional/cross_entropy.py:186]: I am hitting this Assertion also. In my case loss_mask is [1, 768] hence this assertion is hit. Fuses CE implementation seem to expect loss mask to be [768, 1] shape.

**Comment by oleksost** [comment]
[THis](https://github.com/ServiceNow/Fast-LLM/blob/deb7ce66a6bb15e60a493845a71b4f6ee9366ea2/fast_llm/models/gpt/config.py#L182C9-L184C57) should be removed, no?

**Review by oleksost — APPROVED** [review]
Works well for me.
